### PR TITLE
DEP: Deprecate rasterio backend

### LINKED ISSUE #4697: deprecate open_rasterio in favour of a rioxarray entrypoint?
Now that our backends work is well underway, it seems like a good time to discuss deprecating the `open_rasterio` function and outsourcing it to `rioxarray` (which would then implement support for the new entrypoint).

From @jhamman's [comment](https://github.com/pydata/xarray/issues/4142#issuecomment-642413509)

> There's also the question of where the rasterio backend should live once #1970 is implemented. I think we can make a pretty strong argument that it should move out of xarray and into a 3rd party package. Perhaps rasterio itself, or, more likely something like rio-xarray (cc @snowman2). See #3166 for a demo of how this may work.

> As you can see, there are a few big questions up in the air right now. Work on #1970 will begin this summer so now is a great time to game out the various options for the rasterio backend.

I am in favour of doing this. I see no reason why there should be two slightly divergent versions of the same function, and (anecdotally) the `rioxarray` version seems to be more full-featured and regularly improved.

@pydata/xarray What do you think?

If we agree to do this (and rioxarray agrees), we could put this as a "deliverable" in our NASA proposal. This work would be totally relevant to that call.

Related: #4142

cc @snowman2 @scottyhq @JessicaS11 @WeatherGod

### PR DESCRIPTION
**Pull request by snowman2**
- [x] Closes #4697
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by snowman2** [comment]
Fix for doc build issues in #5805

**Comment by keewis** [comment]
thanks, @snowman2, looks good to me. We definitely need a `whats-new.rst` entry for this, though. Once that is done I think we can merge this (unless we're planning to release `0.19.1` first?)

**Comment by snowman2** [comment]
Updated this so it is deprecated in 0.19.1 and added an entry to `whats-new.rst`.

File: doc/whats-new.rst [hunk 1]
```diff
@@ -55,6 +55,8 @@ Breaking changes
 Deprecations
 ~~~~~~~~~~~~
 
+- Deprecate rasterio backend (:issue:`4697`, :pull:`5808`).
```
Comment (keewis) [inline doc/whats-new.rst:58]: If I remember correctly, it was not a backend but just the `xr.open_rasterio` function. If that's correct, I'd change this to:
```suggestion
- Deprecate :py:func:`open_rasterio` (:issue:`4697`, :pull:`5808`).
```

**Review by keewis — COMMENTED** [review]
thanks for the update, @snowman2. I've got some minor suggestions but otherwise looks good to me.

As far as I remember we never used the `deprecated` directive in `xarray` before... it could be useful to add that for other deprecations, too, though.

**Comment by keewis** [comment]
I just pushed a rst comment reminding us to update the deprecation version if we skip 0.19.1, but if I remember the meeting correctly this should now be ready to merge.

Thanks a lot, @snowman2

**Comment by keewis** [comment]
in it goes. Thanks again, @snowman2

**Comment by snowman2** [comment]
Thanks for the review 👍

**Comment by dcherian** [comment]
Thanks @snowman2 This is awesome!

**Pull request merged**